src/storage-client/src/sink.rs

```diff
@@ -282,34 +283,62 @@ pub async fn publish_kafka_schemas(
     topic: String,
     key_schema: Option<String>,
     key_schema_type: Option<mz_ccsr::SchemaType>,
-    value_schema: &str,
+    value_schema: String,
     value_schema_type: mz_ccsr::SchemaType,
+    key_compatibility_level: Option<mz_ccsr::CompatibilityLevel>,
+    value_compatibility_level: Option<mz_ccsr::CompatibilityLevel>,
 ) -> Result<(Option<i32>, i32), anyhow::Error> {
-    let value_schema_id = ccsr
-        .publish_schema(
-            &format!("{}-value", topic),
-            value_schema,
-            value_schema_type,
-            &[],
-        )
+    let key_subject = format!("{}-key", topic);
+    let value_subject = format!("{}-value", topic);
+
+    if let Some(key_compatibility_level) = key_compatibility_level {
+        let ccsr = ccsr.clone();
+        let key_subject = key_subject.clone();
+        async move {
+            ccsr.set_subject_compatibility_level(&key_subject, &key_compatibility_level)
+                .await
+        }
+        .run_in_task(|| "set_key_compatibility_level".to_string())
+        .await
+        .context("unable to update key schema compatibility level in kafka sink")?;
+    }
+
+    if let Some(value_compatibility_level) = value_compatibility_level {
+        let ccsr = ccsr.clone();
+        let value_subject = value_subject.clone();
+        async move {
+            ccsr.set_subject_compatibility_level(&value_subject, &value_compatibility_level)
+                .await
+        }
+        .run_in_task(|| "set_value_compatibility_level".to_string())
         .await
-        .context("unable to publish value schema to registry in kafka sink")?;
+        .context("unable to update value schema compatibility level in kafka sink")?;
+    }
```

Comment:
One thing that worries me is that the CSR options and the topic config options do not function in the same way. The CSR options are force applied repeatedly when the sink starts, while the topic options are only applied once if the topic doesn't exist.

I think we should try to line up the semantics here and try to only set the compatibility level if one hasn't already been configured. I.e., first fetch the existing compatibility level and then only try to set it if it's not already set. There's a small unavoidable race here—someone could sneak in and set it right after we've checked to see that it wasn't set, and before we've had a chance to set it—but I think that is acceptable.

Otherwise, folks might later tweak a compat policy in the CSR directly, only to have Materialize keep overwriting it. And since we don't support `ALTER`ing these properties, so you'd have to drop and recreate the sink to get Materialize to stop doing that.

Reply:
Makes sense to me -- will first check to see if a level has been set before updating it